auto-batching- bare min of the info (#10190)
openvinotoolkit/openvino: docs/OV_Runtime_UG/automatic_batching.md

> # Automatic Batching Plugin {#openvino_docs_OV_UG_Automatic_Batching}

Reviewer: ```suggestion
# Automatic Batching {#openvino_docs_OV_UG_Automatic_Batching}
```

## (Automatic) Batching Execution

The Automatic-Batching is a preview of the new functionality in the OpenVINO™ toolkit.  It performs on-the-fly automatic batching (i.e. grouping inference requests together) to improve device utilization, with no programming effort from the user.
Inputs gathering and outputs scattering from the individual inference requests required for the batch happen transparently, without affecting the application code. 

> The feature primarily targets existing code written for inferencing many requests (each instance with the batch size 1). To obtain corresponding performance improvements, the application must be *running many inference requests simultaneously*. 

Reviewer: ```suggestion

The feature primarily targets existing code written for inferencing many requests (each instance with the batch size 1). To obtain corresponding performance improvements, the application must be *running many inference requests simultaneously*. 
```

The timeout, which adds itself to the execution time of the requests, heavily penalizes the performance. To avoid this, in cases when your parallel slack is bounded, give the OpenVINO an additional hint.

For example, the application processes only 4 video streams, so there is no need to use a batch larger than 4. The most future-proof way to communicate the limitations on the parallelism is to equip the performance hint with the optional `ov::hint::num_requests` configuration key set to *4*. For the GPU this will limit the batch size, for the CPU - the number of inference streams, so each device uses the `ov::hint::num_requests` while converting the hint to the actual device configuration options:
@sphinxdirective

.. tab:: C++

    .. doxygensnippet:: docs/snippets/ov_auto_batching.cpp
       :language: cpp
       :fragment: [hint_num_requests]

.. tab:: Python

    .. doxygensnippet:: docs/snippets/ov_auto_batching.py
       :language: python
       :fragment: hint_num_requests]

@endsphinxdirective

For the *explicit* usage, you can limit the batch size using  "BATCH:GPU(4)",  where 4 is the number of requests running in parallel.

### Other Performance Considerations

> To achieve the best performance with the Automatic Batching it is strongly recommended that the application

Reviewer: ```suggestion
To achieve the best performance with the Automatic Batching, the application should:
```

This _automatic batch size selection_ assumes that the application queries the `ov::optimal_number_of_infer_requests` to create and run the returned number of requests simultaneously:
@sphinxdirective

.. tab:: C++

    .. doxygensnippet:: docs/snippets/ov_auto_batching.cpp
       :language: cpp
       :fragment: [query_optimal_num_requests]

.. tab:: Python

    .. doxygensnippet:: docs/snippets/ov_auto_batching.py
       :language: python
       :fragment: [query_optimal_num_requests]

@endsphinxdirective
If not enough inputs were collected, the `timeout` value makes the transparent execution fall back to the execution of individual requests. Configuration-wise, this is the AUTO_BATCH_TIMEOUT property.
The timeout, which adds itself to the execution time of the requests, heavily penalizes the performance. To avoid this, in cases when your parallel slack is bounded, give the OpenVINO an additional hint.

> For example, the application processes only 4 video streams, so there is no need to use a batch larger than 4. The most future-proof way to communicate the limitations on the parallelism is to equip the performance hint with the optional `ov::hint::num_requests` configuration key set to 4. For the GPU this will limit the batch size, for the CPU - the number of inference streams, so each device uses the `ov::hint::num_requests` while converting the hint to the actual device configuration options:

Reviewer 2: `For the GPU this will limit the batch size, for the CPU - the number of inference streams` - is it important for me? 
Guide should be device agnostic, while you are trying to provide complex details which are not needed for me

Reviewer: > `For the GPU this will limit the batch size, for the CPU - the number of inference streams` - is it important for me? Guide should be device agnostic, while you are trying to provide complex details which are not needed for me

Merging the PR as is. The comments will be addressed later in a separate PR with fixes.

# Automatic Batching Plugin {#openvino_docs_OV_UG_Automatic_Batching}

## (Automatic) Batching Plugin Execution (C++)

@sphinxdirective
.. raw:: html

    <div id="switcher-cpp" class="switcher-anchor">C++</div>

> @endsphinxdirective

Reviewer 2: I suppose the content of your page is mostly language independent. Do we need separation between C++ and python?

Author: the new config names are C++ e.g. ov::hint::performance_mode (shall we put the string names to accomodate the Python)

> @sphinxdirective
> .. raw:: html
> 
>     <div id="switcher-cpp" class="switcher-anchor">C++</div>
> @endsphinxdirective

Reviewer: ```suggestion
```

> The BATCH device is a preview of the new, special "virtual" device in the OpenVINO™ toolkit. It performs on-the-fly automatic batching (i.e. grouping inference requests together) to improve device utilization, with no programming effort from the user side.

Reviewer 2: `BATCH device` => `auto-batch execution` ?

> The BATCH device is a preview of the new, special "virtual" device in the OpenVINO™ toolkit. It performs on-the-fly automatic batching (i.e. grouping inference requests together) to improve device utilization, with no programming effort from the user side.

Author: utilization == how efficient is the usage

Gathering of the inputs (as well as scattering the outputs) from the individual inference requests that is needed to constitute the batch happens in the completely transparent way, so the application code is left intact. The major pre-requisite for the application to enjoy an associated performance improvements is *running many inference requests simultaneously*.
So the feature primarily targets existing codes that are written for inferencing many requests (each instances with batch size of 1).
  

> Since the batching is straightforward way of leveraging the GPU compute power (and save on communication overheads) the automatic batching is _implicitly_ triggered on the GPU when the ov::hint::PerformanceMode::THROUGHPUT is specified for the ov::hint::performance_mode property for the compile_model or set_property calls. 

Reviewer 2: compile_model => `compile_model ` ?

> **NOTE**: You can disable the Auto-Batching (e.g. for the GPU device) from being triggered by the ov::hint::PerformanceMode::THROUGHPUT. Just pass the ov::hint::allow_auto_batching (set to the 'false') in addition to the ov::hint::performance_mode .

Reviewer 2: please, provide working code snippet how to do it

Alternatively (in the legacy apps that are not akin to the notion of the performance hints) you may want to use the explicit device notion like 'BATCH:GPU'. In both cases the actual batch size selection happens on the per-model basis (also respecting things like on-device memory for the dGPUs).

Notice that _automatic batch size selection_ assumes that the application queries the ov::optimal_number_of_infer_requests to create and runs simultaneously the returned number of requests. This is very important, as there is a "timeout" value so that (when no enough inputs have been collected) the execution transparently falls back to execution the requests individually. Configuration-wise this is AUTO_BATCH_TIMEOUT property, please see below.
The timeout, which adds itself to the execution time of the requests, heavily penalizes the performance. To avoid it, in the cases when your parallel slack is bounded (e.g. the application processes only 4 video-streams, so batch larger than 4 doesn't make sense) make sure to give the OpenVINO an additional hint the OV on this.

> When the performance hint is equipped with the optional ov::hint::num_requests configuration key set to "4". This is the most future-proof way to communicate the limitations on the parallelism. E.g. for the GPU that will limit the batch size, for the CPU, the number of inference streams <link> accordingly, so each device uses the ov::hint::num_requests when converting the hint to the actual device configuration options.

Reviewer: <link> - please, specify the link

> For the explicit usage i.e. the BATCH device you can limit the batch size via "BATCH:GPU(4)",  when you have only 4 requests running in parallel.

Reviewer: ```suggestion
For the **explicit ** usage, you can limit the batch size using "BATCH:GPU(4)",  where 4 is the number of requests running in parallel.
```

Limitations:

>  - Although less critical for the throughput-oriented scenarios, the load-time is increased almost 2x when auto-batching is used 

Reviewer: ```suggestion
 - Although less critical for the throughput-oriented scenarios, the load-time with auto-batching increases by almost 2x 
```

>  - Certain networks are not reshape-able by the batching (aka 'N', using the layouts parlance) dimension, so the auto-batching will not be triggered 

Reviewer: ```suggestion
 - Certain networks are not reshape-able by the batching dimension (specified as 'N' in the layouts terms), so the auto-batching is not triggered 
```

>  - Memory footpint growth is unavoidable price. Yet the auto-batching queries the available memory (especially for the dGPUs) and limits the selected batch size accordingly.

Reviewer: ```suggestion
- Although performance improvements happen at the cost of the memory footprint growth, the auto-batching queries the available memory (especially for the dGPUs) and limits the selected batch size accordingly
```

 

### Configuring the Automatic Batching Plugin

> Following the OpenVINO convention for devices names, the  "batching" device is named as  "BATCH". The configuration options for that are as follows:

Reviewer: ```suggestion
Following the OpenVINO convention for devices names, the  *batching* device is named  *BATCH*. The configuration options are as follows:
```

| Parameter name     | Parameter description      | Default            |             Examples                                                      |
| :---               | :---                  | :---               |:-----------------------------------------------------------------------------|
| "AUTO_BATCH_DEVICE" | device name to apply the automatic batching and optional batch size in brackets | N/A | BATCH:GPU which triggers the automatic batch size selection or explicit batch size BATCH:GPU(4)     |

> | "AUTO_BATCH_TIMEOUT" | timeout value, in ms | 1000 |  you can reduce the timeout value (to avoid performance penalty when the data arrives too non-evenly) e.g. pass the "100", or in contrast make it large enough e.g. to accommodate inputs preparation (e.g. when it is serial process)     |

Reviewer 2: use new configuration properties

Author: will do a seperate PR

> | "AUTO_BATCH_TIMEOUT" | timeout value, in ms | 1000 |  you can reduce the timeout value (to avoid performance penalty when the data arrives too non-evenly) e.g. pass the "100", or in contrast make it large enough e.g. to accommodate inputs preparation (e.g. when it is serial process)     |

Reviewer: ```suggestion
| "AUTO_BATCH_TIMEOUT" | timeout value, ms | 1000 |  User can reduce the timeout value to avoid performance penalty when the data arrives non-evenly, for example, by passing *100*. Or, on the contrary, make it large enough to accommodate the preparation of inputs, when it happens as a serial process. |
```

# Automatic Batching Plugin {#openvino_docs_OV_UG_Automatic_Batching}

## (Automatic) Batching Execution

> The Automatic-Batching is a preview of the new functionality in the OpenVINO™ toolkit.  It performs on-the-fly automatic batching (i.e. grouping inference requests together) to improve device utilization, with no programming effort from the user.

Author: again, usage is not the same as utilization

For the **explicit** usage, you can limit the batch size using  "BATCH:GPU(4)",  where 4 is number of requests running in parallel.

Limitations:
 - Although less critical for the throughput-oriented scenarios, the load-time with auto-batching increases by almost 2x.
 - Certain networks are not reshape-able by the batching dimension (specified as 'N' in the layouts terms), so the auto-batching is not triggered.
 - Performance improvements happen at the cost of the memory footprint growth, yet the auto-batching queries the available memory (especially for the dGPUs) and limits the selected batch size accordingly.

 

### Configuring the Automatic Batching Plugin
Following the OpenVINO convention for devices names, the *batching* device is named *BATCH*. The configuration options are as follows:

| Parameter name     | Parameter description      | Default            |             Examples                                                      |
| :---               | :---                  | :---               |:-----------------------------------------------------------------------------|

> | "AUTO_BATCH_DEVICE" | device name to apply the automatic batching and optional batch size in brackets | N/A | BATCH:GPU which triggers the automatic batch size selection or explicit batch size BATCH:GPU(4)     |

Reviewer: ```suggestion
| "AUTO_BATCH_DEVICE" | Device name to apply the automatic batching and optional batch size in brackets | N/A | BATCH:GPU which triggers the automatic batch size selection or explicit batch size BATCH:GPU(4)     |
```

Inputs gathering and outputs scattering from the individual inference requests required for the batch happen transparently, without affecting the application code. 
The feature primarily targets existing code written for inferencing many requests (each instance with the batch size 1). To obtain corresponding performance improvements, the application must be *running many inference requests simultaneously*. 
As explained below, the auto-batching functionality can be also used via a special *virtual* device.       

Batching is a straightforward way of leveraging the GPU compute power and saving on communication overheads. The automatic batching is  _implicitly_ triggered on the GPU when the `ov::hint::PerformanceMode::THROUGHPUT` is specified for the `ov::hint::performance_mode` property for the compile_model or set_property calls.
@sphinxdirective

.. tab:: C++

    .. doxygensnippet:: docs/snippets/ov_auto_batching.cpp
       :language: cpp
       :fragment: [compile_model]

.. tab:: Python

    .. doxygensnippet:: docs/snippets/ov_auto_batching.py
       :language: python
       :fragment: [compile_model]

@endsphinxdirective

> > **NOTE**: You can disable the Auto-Batching (for example, for the GPU device) from being triggered by the `ov::hint::PerformanceMode::THROUGHPUT`. To do that, pass the `ov::hint::allow_auto_batching` set to 'false' in addition to the `ov::hint::performance_mode`:

Reviewer: ```suggestion
> **NOTE**: You can disable the Auto-Batching (for example, for the GPU device) from being triggered by the `ov::hint::PerformanceMode::THROUGHPUT`. To do that, pass the `ov::hint::allow_auto_batching` set to **false** in addition to the `ov::hint::performance_mode`:
```

@sphinxdirective

.. tab:: C++

    .. doxygensnippet:: docs/snippets/ov_auto_batching.cpp
       :language: cpp
       :fragment: [compile_model_no_auto_batching]

.. tab:: Python

    .. doxygensnippet:: docs/snippets/ov_auto_batching.py
       :language: python
       :fragment: [compile_model_no_auto_batching]

@endsphinxdirective

Alternatively, to enable the Auto-Batching in the legacy apps not akin to the notion of the performance hints, you may need to use the **explicit** device notion, such as 'BATCH:GPU'. In both cases, the batch size selection depends on a model and device specifics (for example, on-device memory for the dGPUs).

The _automatic batch size selection_ assumes that the application queries the `ov::optimal_number_of_infer_requests` to create and run the returned number of requests simultaneously. If not enough inputs were collected, the `timeout` value makes the transparent execution fall back to the execution of individual requests. Configuration-wise, this is the AUTO_BATCH_TIMEOUT property. 

> The timeout, which adds itself to the execution time of the requests, heavily penalizes the performance. To avoid this, in cases when your parallel slack is bounded, (for example, the application processes only 4 video streams, so there is no need to use a batch larger than 4) give the OpenVINO an additional hint.
> So the performance hint should be equipped with the optional `ov::hint::num_requests` configuration key set to **4**. 
> 
> This is the most future-proof way to communicate the limitations of parallelism. For example, for the GPU it limits the batch size, for the CPU - the number of inference streams. Each device uses `the ov::hint::num_requests` when converting the hint to the device configuration options.

Reviewer: ```suggestion
The timeout, which adds itself to the execution time of the requests, heavily penalizes the performance. To avoid this, in cases when your parallel slack is bounded, give the OpenVINO an additional hint.

For example, the application processes only 4 video streams, so there is no need to use a batch larger than 4. The most future-proof way to communicate the limitations on the parallelism is to equip the performance hint with the optional `ov::hint::num_requests` configuration key set to *4*. For the GPU this will limit the batch size, for the CPU - the number of inference streams, so each device uses the `ov::hint::num_requests` while converting the hint to the actual device configuration options.
```

> For the **explicit** usage, you can limit the batch size using  "BATCH:GPU(4)",  where 4 is number of requests running in parallel.

Reviewer: ```suggestion
For the *explicit* usage, you can limit the batch size using  "BATCH:GPU(4)",  where 4 is the number of requests running in parallel.
```
